### PR TITLE
[Refactor] Fixing mypy issues of `service/user`

### PR DESCRIPTION
**Pull request by khoaguin**

### CODE REVIEW

File: packages/syft/src/syft/service/user/user_roles.py [hunk 1]
```diff
@@ -36,14 +37,14 @@ class ServiceRole(Enum):
     # Disabling it, as both property and classmethod only works for python >= 3.9
     # @property
     @classmethod
-    def roles_descending(cls) -> List[Tuple[int, Self]]:
+    def roles_descending(cls) -> List[Tuple[int, "ServiceRole"]]:
```
Comment (kiendang) [inline packages/syft/src/syft/service/user/user_roles.py]: Shouldn't `Self` still work for `@classmethod`?

File: packages/syft/src/syft/service/user/user_roles.py [hunk 2]
```diff
@@ -36,14 +37,14 @@ class ServiceRole(Enum):
...
         tuples = []
         for x in cls:
             tuples.append((x.value, x))
         return sorted(tuples, reverse=True)
 
     @staticmethod
-    def roles_for_level(level: Union[int, Self]) -> List[Self]:
+    def roles_for_level(level: Union[int, "ServiceRole"]) -> List["ServiceRole"]:
```
Comment (kiendang) [inline packages/syft/src/syft/service/user/user_roles.py]: This should be a `@classmethod` instead.
Then line 52 `service_roles = ServiceRole.roles_descending()` should be changed to `service_roles = cls.roles_descending()`

File: packages/syft/src/syft/service/user/user_roles.py [hunk 3]
```diff
@@ -90,19 +91,21 @@ def __lt__(self, other: Self) -> bool:
...
 
-DATA_OWNER_ROLE_LEVEL = ServiceRole.roles_for_level(
+DATA_OWNER_ROLE_LEVEL: List[ServiceRole] = ServiceRole.roles_for_level(
     ServiceRole.DATA_OWNER + ServiceRole.ADMIN
 )
 
 ADMIN_ROLE_LEVEL = ServiceRole.roles_for_level(ServiceRole.ADMIN)
 
-ROLE_TO_CAPABILITIES = {
+ROLE_TO_CAPABILITIES: Dict["ServiceRole", List["ServiceRoleCapability"]] = {
```
Comment (kiendang) [inline packages/syft/src/syft/service/user/user_roles.py]: This should work without quotes?

File: packages/syft/src/syft/service/user/user_service.py [hunk 4]
```diff
@@ -180,9 +182,9 @@ def search(
             ]
 
             # Return the proper slice using page_index
-            results = results[page_index]
-
-            results = UserViewPage(users=results, total=total)
+            if page_index:
```
Comment (kiendang) [inline packages/syft/src/syft/service/user/user_service.py]: `page_index` could be 0 which would be falsy here. Better to check the logic and see whether `if page_index is None` is more appropriate.

Comment (khoaguin) [reply packages/syft/src/syft/service/user/user_service.py]: Fixed. Thanks @kiendang

File: packages/syft/src/syft/service/user/user_service.py [hunk 5]
```diff
@@ -317,7 +319,9 @@ def update(
 
         return user.to(UserView)
 
-    def get_target_object(self, credentials: SyftVerifyKey, uid: UID):
+    def get_target_object(
+        self, credentials: SyftVerifyKey, uid: UID
+    ) -> Union[SyftError, User]:
```
Comment (kiendang) [inline packages/syft/src/syft/service/user/user_service.py]: nitpick: usually the convention is that the error type is on the right and the success type is on the left.

```suggestion
    ) -> Union[User, SyftError]:
```

**Review by kiendang — APPROVED** [review]
Great work @khoaguin

**Review by kiendang — APPROVED** [review]

File: packages/syft/src/syft/service/user/user_roles.py [hunk 6]
```diff
@@ -59,7 +60,7 @@ def roles_for_level(level: Union[int, Self]) -> List[Self]:
                 level_float = level_float % role_num
         return roles
 
-    def capabilities(self):
+    def capabilities(self) -> List["ServiceRoleCapability"]:
```
Comment (kiendang) [inline packages/syft/src/syft/service/user/user_roles.py]:
```suggestion
    def capabilities(self) -> List[ServiceRoleCapability]:
```

File: .pre-commit-config.yaml [hunk 7]
```diff
@@ -172,7 +172,7 @@ repos:
       - id: mypy
         name: "mypy: syft"
         always_run: true
-        files: "^packages/syft/src/syft/serde|^packages/syft/src/syft/util/env.py|^packages/syft/src/syft/util/logger.py|^packages/syft/src/syft/util/markdown.py|^packages/syft/src/syft/util/notebook_ui/notebook_addons.py|^packages/syft/src/syft/service/warnings.py|^packages/syft/src/syft/service/dataset|^packages/syft/src/syft/service/worker"
+        files: "^packages/syft/src/syft/serde|^packages/syft/src/syft/util/env.py|^packages/syft/src/syft/util/logger.py|^packages/syft/src/syft/util/markdown.py|^packages/syft/src/syft/util/notebook_ui/notebook_addons.py|^packages/syft/src/syft/service/warnings.py|^packages/syft/src/syft/service/dataset|^packages/syft/src/syft/service/worker||^packages/syft/src/syft/service/user"
```
Comment (kiendang) [inline .pre-commit-config.yaml]:
```suggestion
        files: "^packages/syft/src/syft/serde|^packages/syft/src/syft/util/env.py|^packages/syft/src/syft/util/logger.py|^packages/syft/src/syft/util/markdown.py|^packages/syft/src/syft/util/notebook_ui/notebook_addons.py|^packages/syft/src/syft/service/warnings.py|^packages/syft/src/syft/service/dataset|^packages/syft/src/syft/service/worker|^packages/syft/src/syft/service/user"
```

**Pull request merged**